### PR TITLE
BUG: transpose casts mixed dtypes to object

### LINKED ISSUE #43337: BUG: transpose unnecessarily returning object
Ref: https://github.com/pandas-dev/pandas/pull/43213/files#r699680952

Example:

```
df = pd.DataFrame({'a': [1], 'b': [2]})
print(df.T.dtypes)
df['b'] = df['b'].astype('Int64')
print(df.T.dtypes)
```

gives int64 for the first example, and object for the 2nd. However, the 2nd arguably would be better off as Int64.

The code from #43213 could be used in the transpose method, something akin to

```
from pandas.core.dtypes.cast import find_common_type

df = pd.DataFrame({'a': [1], 'b': [2]})
df['b'] = df['b'].astype('Int64')
common_type = find_common_type(df.dtypes.tolist())
print(df.T.astype(common_type, copy=False).dtypes)
```

Perhaps there are undesirable side-effects by doing this?

### PR DESCRIPTION
**Pull request by debnathshoham**
- [x] closes #43337
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by debnathshoham** [comment]
All the failures are from the below two, not sure why .These are passing in some cases (also in my local). 
Would be really helpful if someone can point me to the correct direction.
```
FAILED pandas/tests/frame/methods/test_transpose.py::TestTranspose::test_transpose_tzaware_2col_mixed_tz
FAILED pandas/tests/frame/methods/test_transpose.py::TestTranspose::test_transpose_object_to_tzaware_mixed_tz

**Comment by mzeitlin11** [comment]
> All the failures are from the below two, not sure why .These are passing in some cases (also in my local).
> Would be really helpful if someone can point me to the correct direction.
> 
> ```
> FAILED pandas/tests/frame/methods/test_transpose.py::TestTranspose::test_transpose_tzaware_2col_mixed_tz
> FAILED pandas/tests/frame/methods/test_transpose.py::TestTranspose::test_transpose_object_to_tzaware_mixed_tz
> ```

Based on a quick look, initial guess might be different `numpy` versions. Maybe try using the same `numpy` version as one of the failing builds (or just see if you can reproduce with an matching requirements file, which you can find at `pandas/ci/deps`)

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3381,7 +3381,12 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
             new_arr = self.values.T
             if copy:
                 new_arr = new_arr.copy()
-            result = self._constructor(new_arr, index=self.columns, columns=self.index)
+            result = self._constructor(
+                new_arr,
+                index=self.columns,
+                columns=self.index,
+                dtype=find_common_type(dtypes) if len(dtypes) > 0 else None,
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3388]: can you put this on its own line

File: pandas/core/frame.py [hunk 2]
```diff
@@ -3381,7 +3381,12 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
             new_arr = self.values.T
             if copy:
                 new_arr = new_arr.copy()
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3383]: copy here might be unnecessary if we end up casting below

**Comment by jbrockmendel** [comment]
> test_transpose_tzaware_2col_mixed_tz

Best guess is something weird is going on with find_common_type between dt64tz[US/Eastern] and dt64tz[US/Pacific].  I think find_common_type should return object, but in some places we pretend that the common type is dt64tz[UTC]

**Comment by jreback** [comment]
ok if you can merge master will look again

File: pandas/tests/frame/methods/test_transpose.py [hunk 3]
```diff
@@ -57,6 +58,7 @@ def test_transpose_object_to_tzaware_mixed_tz(self):
         df2 = DataFrame([dti, dti2])
         assert (df2.dtypes == object).all()
         res2 = df2.T
+        print("\n", res2.dtypes, [dti.dtype, dti2.dtype])
```
Comment (jreback) [inline pandas/tests/frame/methods/test_transpose.py:61]: remove prints

Comment (debnathshoham) [reply pandas/tests/frame/methods/test_transpose.py]: removed the prints
these are the only two tests failing (but not on my local, was trying to figure why)

File: doc/source/whatsnew/v1.4.0.rst [hunk 4]
```diff
@@ -501,6 +501,7 @@ Reshaping
 - Improved error message when creating a :class:`DataFrame` column from a multi-dimensional :class:`numpy.ndarray` (:issue:`42463`)
 - :func:`concat` creating :class:`MultiIndex` with duplicate level entries when concatenating a :class:`DataFrame` with duplicates in :class:`Index` and multiple keys (:issue:`42651`)
 - Bug in :meth:`pandas.cut` on :class:`Series` with duplicate indices (:issue:`42185`) and non-exact :meth:`pandas.CategoricalIndex` (:issue:`42425`)
+- Bug in :meth:`DataFrame.transpose`, where mixed dtypes were cast to ``object`` (:issue:`43337`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:504]: this is saying the opposite of what you are doing. make this more clear to a reader

File: pandas/core/frame.py [hunk 5]
```diff
@@ -3400,9 +3400,13 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
 
         else:
             new_arr = self.values.T
-            if copy:
-                new_arr = new_arr.copy()
-            result = self._constructor(new_arr, index=self.columns, columns=self.index)
+            common_dtype = find_common_type(dtypes) if len(dtypes) > 0 else None
+            result = self._constructor(
```
Comment (jreback) [inline pandas/core/frame.py:3404]: pass copy to the constructor (and are you testing this)?

File: pandas/tests/frame/methods/test_transpose.py [hunk 6]
```diff
@@ -103,3 +103,10 @@ def test_transpose_get_view_dt64tzget_view(self):
 
         rtrip = result._mgr.blocks[0].values
         assert np.shares_memory(arr._data, rtrip._data)
+
+    def test_transpose_mixed_dtypes(self):
```
Comment (jreback) [inline pandas/tests/frame/methods/test_transpose.py:107]: test the copy keyword. test an example which does keep object (e.g. (1, 1.5) and ( 'a', 1)) for (a, b)  and test with simple dtypes as well (e.g. int64)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
this very likley changes a number of tests you will have to edit those

**Comment by jreback** [comment]
status of this?

**Comment by debnathshoham** [comment]
sorry, was not able to give time into this.. 
will take a look again, sorry for the delay

**Comment by mroeschke** [comment]
Thanks @debnathshoham but appears this PR has gone stale. When you have the time feel free to merge master and we can reopen. Closing for now.

**Pull request closed without merging**